Patching an __init__ that calls super() doesn't work

It fails with

```
  File "/opt/yplan/venv3/lib/python3.5/site-packages/patchy/api.py", line 174, in _set_source
    real_func.__code__ = new_code
ValueError: __init__() requires a code object with 1 free vars, not 0
```

From reading here:
http://stackoverflow.com/questions/4885168/how-to-make-super-work-by-manually-filling-the-class-cell

My general understanding is that to make `super()` work on Py3 they turned all `__init__`'s into closures which source `__class__` from the containing scope (regardless of whether the `__init__` actually uses `super`).
Patchy doesn't currently support closures (in general how would you access them to patch them) and so doesn't preserve the `co_freevars` on the code object.

This isn't Python 3 only, it fails on Python 2 as well with the same error.

I've tried it within the patchy test suite with this test:

```python

    def test_patch_init_super(self):
        class Person(object):
            def __init__(self):
                pass

        class Artist(Person):
            def __init__(self):
                super(Person, self).__init__()
                self.prop = 'old'

        patchy.patch(Artist.__init__, """\
            @@ -1,3 +1,3 @@
             def __init__(self):
                 super().__init__()
            -    self.prop = 'old'
            +    self.prop = 'new'""")

        a = Artist()
        assert a.prop == 'new'
```

Which yields this result:

```
==================================================================================== FAILURES ====================================================================================
________________________________________________________________________ PatchTests.test_patch_init_super ________________________________________________________________________

self = <tests.test_patchy.PatchTests testMethod=test_patch_init_super>

    def test_patch_init_super(self):
        class Person(object):
            def __init__(self):
                pass

        class Artist(Person):
            def __init__(self):
                super(Person, self).__init__()
                self.prop = 'old'

        patchy.patch(Artist.__init__, """\
                @@ -1,3 +1,3 @@
                 def __init__(self):
                     super().__init__()
                -    self.prop = 'old'
>               +    self.prop = 'new'""")

tests/test_patchy.py:232:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
patchy/api.py:23: in patch
    return _do_patch(func, patch_text, forwards=True)
patchy/api.py:60: in _do_patch
    _set_source(func, new_source)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

func = <unbound method Artist.__init__>, new_source = "def __init__(self):\n    super(Person, self).__init__()\n    self.prop = 'new'\n"

    def _set_source(func, new_source):
        # Fetch the actual function we are changing
        real_func = _get_real_func(func)

        # Figure out any future headers that may be required
        feature_flags = real_func.__code__.co_flags & FEATURE_MASK

        # Compile and retrieve the new Code object
        localz = {}
        new_code = compile(
            new_source,
            '<patchy>',
            'exec',
            flags=feature_flags,
            dont_inherit=True
        )

        six.exec_(new_code, func.__globals__, localz)
        new_func = localz[func.__name__]

        # Figure out how to get the Code object
        if isinstance(new_func, (classmethod, staticmethod)):
            new_code = new_func.__func__.__code__
        else:
            new_code = new_func.__code__

        # Put the new Code object in place
>       real_func.__code__ = new_code
E       ValueError: __init__() requires a code object with 1 free vars, not 0
```

There are some hints [on SO](https://stackoverflow.com/questions/4885168/how-to-make-super-work-by-manually-filling-the-class-cell) about this, but the main answer there involves replacing the function object rather than just the code object, and patchy at current restricts its fiddling to the code object and it would be nice to keep to that. Perhaps there is a way to do this, unclear right now.

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -8,6 +8,9 @@ Pending Release
 
 .. Insert new release notes below this line
 
+* Support cases where the function or class name is not in the freevars. This
+  allows patching ``__init__`` of a class, or patching a recursive module-level function.
+
 1.4.0 (2017-05-13)
 ------------------
 

--- a/patchy/api.py
+++ b/patchy/api.py
@@ -245,7 +245,14 @@ def _set_source(func, func_source):
         """
 
         _def, _ast, fv_body = _process_freevars()
-        class_src = '    class {name}(object):\n        pass'.format(name=class_name)
+        _global = (
+            '' if class_name in func.__code__.co_freevars
+            else '    global {name}\n'.format(name=class_name)
+        )
+        class_src = '{_global}    class {name}(object):\n        pass'.format(
+            _global=_global,
+            name=class_name,
+        )
         ret = '    return {class_name}.{name}'.format(
             class_name=class_name,
             name=func.__name__,
@@ -272,7 +279,7 @@ def _set_source(func, func_source):
     localz = {}
     new_code = _compile(new_source)
 
-    six.exec_(new_code, func.__globals__, localz)
+    six.exec_(new_code, dict(func.__globals__), localz)
     new_func = localz['__patchy_freevars__']()
 
     # Figure out how to get the Code object

--- a/patchy/api.py
+++ b/patchy/api.py
@@ -264,8 +264,13 @@ def _set_source(func, func_source):
 
     def _process_function():
         _def, _ast, fv_body = _process_freevars()
-        ret = '    return {name}'.format(name=func.__name__)
-        to_parse = '\n'.join([_def] + fv_body + ['    pass', ret])
+        name = func.__name__
+        ret = '    return {name}'.format(name=name)
+        _global = (
+            [] if name in func.__code__.co_freevars
+            else ['    global {name}'.format(name=name)]
+        )
+        to_parse = '\n'.join([_def] + _global + fv_body + ['    pass', ret])
         new_source = _parse(to_parse)
         new_source.body[0].body[-2] = _ast
         return new_source

--- a/tests/test_patch.py
+++ b/tests/test_patch.py
@@ -208,7 +208,7 @@ def test_patch_init_module_level(tmpdir):
     Module level classes do not have a freevar for their class name, whilst
     classes defined in a scope do...
     """
-    example_py = tmpdir.join('example.py')
+    example_py = tmpdir.join('patch_init_module_level.py')
     example_py.write(dedent('''\
         class Person(object):
             def __init__(self):
@@ -235,6 +235,40 @@ def test_patch_init_module_level(tmpdir):
     assert mod.Artist == Artist
     assert a.base_prop == 'yo'
     assert a.prop == 'new'
+
+
+def test_patch_recursive_module_level(tmpdir):
+    """
+    Module level recursive functions do not have a freevar for their name,
+    whilst functions defined in a scope do...
+    """
+    example_py = tmpdir.join('patch_recursive_module_level.py')
+    example_py.write(dedent("""\
+        def factorial(n):
+            if n == 1:
+                return 1
+            else:
+                return n * factorial(n-1)
+    """))
+    mod = example_py.pyimport()
+    factorial = mod.factorial
+
+    assert factorial(10) == 3628800
+
+    patchy.patch(factorial, """\
+        @@ -2,4 +2,4 @@
+        -    if n == 1:
+        -        return 1
+        -    else:
+        -        return n * factorial(n-1)
+        +   if n <= 1:
+        +       return n
+        +   else:
+        +       return factorial(n-1) + factorial(n-2)
+    """)
+
+    assert factorial(10) == 55
+    assert factorial == mod.factorial
 
 
 @skip_unless_python_3  # PEP 3135 New Super

--- a/tests/test_patch.py
+++ b/tests/test_patch.py
@@ -203,7 +203,6 @@ def test_patch_instancemethod_freevars():
     assert Artist().method() == "Cheese on toast"
 
 
-@pytest.mark.xfail(raises=ValueError)
 def test_patch_init_module_level(tmpdir):
     """
     Module level classes do not have a freevar for their class name, whilst
@@ -233,6 +232,7 @@ def test_patch_init_module_level(tmpdir):
     """)
 
     a = Artist()
+    assert mod.Artist == Artist
     assert a.base_prop == 'yo'
     assert a.prop == 'new'
 

